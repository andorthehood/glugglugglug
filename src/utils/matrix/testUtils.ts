export function roundValue(value: number, precision = 6) {
	const factor = 10 ** precision;
	const rounded = Math.round(value * factor) / factor;
	return Object.is(rounded, -0) ? 0 : rounded;
}

export function formatMatrix(values: number[]) {
	return Array.from({ length: 4 }, (_, row) => {
		const start = row * 4;
		return values.slice(start, start + 4).map((value) => roundValue(value));
	});
}

export function formatVector(values: number[]) {
	return values.map((value) => roundValue(value));
}
