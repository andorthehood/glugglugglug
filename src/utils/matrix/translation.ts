export function translation(tx: number, ty: number, tz: number) {
	return [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, tx, ty, tz, 1];
}
