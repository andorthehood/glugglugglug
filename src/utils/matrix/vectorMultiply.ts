export function vectorMultiply(v: number[], m: number[]) {
	const dst = [];
	for (let i = 0; i < 4; ++i) {
		dst[i] = 0.0;
		for (let j = 0; j < 4; ++j) {
			dst[i] += v[j] * m[j * 4 + i];
		}
	}
	return dst;
}
