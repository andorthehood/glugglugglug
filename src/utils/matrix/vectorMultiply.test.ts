import { describe, expect, it } from "vitest";
import { vectorMultiply } from "./vectorMultiply";
import { translation } from "./translation";
import { formatVector } from "./testUtils";

describe("vectorMultiply", () => {
	it("multiplies a vector by a matrix", () => {
		const vec = [1, 2, 3, 1];
		const mat = translation(4, -5, 6);

		expect(formatVector(vectorMultiply(vec, mat))).toMatchSnapshot();
	});
});
