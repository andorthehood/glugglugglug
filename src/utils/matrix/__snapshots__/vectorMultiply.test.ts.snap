// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vectorMultiply > multiplies a vector by a matrix 1`] = `
[
  5,
  -3,
  9,
  1,
]
`;
